src/levanter/layers/gated_deltanet.py

```diff
@@ -0,0 +1,851 @@
+# - the JAX implementation by Yu Sun and Leo Lee
+
+"""
+This module implements the Gated DeltaNet: https://arxiv.org/abs/2412.06464.
+It exposes:
+  - recurrent_gated_delta_rule: the sequential (decode) rule
+  - chunk_gated_delta_rule:    the chunkwise-parallel (prefill / train) rule
+  - GatedDeltaNet:             a full layer that wraps projections, a small
+                               depthwise causal conv over [Q|K|V], the kernels,
+                               and the gated RMSNorm + output projection.
+
+Core update (rectangular state S ∈ R^{d_k × d_v}):
+  S_t = α_t S_{t-1} + β_t (v_t - S_{t-1} k_t) k_t^T
+  o_t = S_t^T q_t
+
+where:
+  α_t = exp(g_t) ∈ (0,1)   (forget/decay gate, log-parameterized by g_t ≤ 0)
+  β_t = σ(b_t) ∈ (0,1)     (learning-rate gate)
+
+Follows the GDN implementation for Qwen3-Next. Notably most math is performed in fp32.
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+from jax import lax
+
+import haliax as hax
+import haliax.nn as hnn
+from haliax import Axis, NamedArray
+
+
+# ---------- small utilities ----------
+
+
+def _l2norm(x: NamedArray, axis: hax.AxisSelector, eps: float = 1e-6) -> NamedArray:
+    """L2-normalize x along a named axis.
+
+    Args:
+        x: NamedArray of any shape.
+        axis: the single axis to normalize along (e.g., the head dimension Dk).
+    """
+    x32 = x.astype(jnp.float32)
+    inv = hax.rsqrt(hax.sum(hax.square(x32), axis=axis) + jnp.asarray(eps, dtype=jnp.float32))
+    return (x32 * inv).astype(x.dtype)
+
+
+# ---------- depthwise conv: positional (lax) helpers with named wrappers ----------
+
+
+def _causal_depthwise_conv1d_full(
+    x_ncl: jnp.ndarray, w_ck: jnp.ndarray, bias_c: Optional[jnp.ndarray] = None
+) -> jnp.ndarray:
+    """Depthwise 1D convolution with *causal* semantics (left padding).
+
+    Shapes:
+      x_ncl: (N, C, L)  - batch, channels, length
+      w_ck:  (C, K)     - per-channel (depthwise) filter of length K
+      bias:  (C,)       - optional per-channel bias
+      return: (N, C, L)
+
+    DimensionNumbers ("NCH","OIH","NCH") means:
+    - lhs (x):    N=0, C=1, H=2
+    - rhs (w):    O=0, I=1, H=2  (we inject a singleton I=1 for depthwise)
+    - out:        N=0, C=1, H=2
+    """
+    N, C, L = x_ncl.shape
+    K = w_ck.shape[-1]
+    # pad x on the left with K-1 zeros so that output length == L ("causal")
+    x_pad = jnp.pad(x_ncl, ((0, 0), (0, 0), (K - 1, 0)))
+    w_oik = w_ck[:, None, :]  # (C, 1, K) → O=C, I=1, K
+    y = lax.conv_general_dilated(
+        lhs=x_pad,
+        rhs=w_oik,
+        window_strides=(1,),
+        padding="VALID",
+        dimension_numbers=("NCH", "OIH", "NCH"),
+        feature_group_count=C,  # depthwise
+        preferred_element_type=jnp.float32,
+    )
+    if bias_c is not None:
+        y = y + bias_c[:, None]
+    y = jax.nn.silu(y)
+    return y
```

Comment:
after this pr, i'd spend some time making sure the shardings are what you expect (jax.debug.inspect_sharding iirc)